File: epoxy-adapter/src/main/java/com/airbnb/epoxy/ControllerModelList.java
package com.airbnb.epoxy;

/**
 * This ArrayList subclass enforces that no changes are made to the list after {@link #freeze()} is
 * called. This prevents model interceptors from storing the list and trying to change it later. We
 * could copy the list before diffing, but that would waste memory to make the copy for every
 * buildModels cycle, plus the interceptors could still try to modify the list and be confused about
 * why it doesn't do anything.
 */
class ControllerModelList extends ModelList {

  private static final ModelListObserver OBSERVER = new ModelListObserver() {
    @Override
    public void onItemRangeInserted(int positionStart, int itemCount) {
      throw new IllegalStateException(
          "Models cannot be changed once they are added to the controller");
    }

    @Override
    public void onItemRangeRemoved(int positionStart, int itemCount) {
      throw new IllegalStateException(
          "Models cannot be changed once they are added to the controller");
    }
  };

  ControllerModelList(int expectedModelCount) {
    super(expectedModelCount);
    pauseNotifications();
  }

  void freeze() {
    setObserver(OBSERVER);
    resumeNotifications();
  }
}

File: epoxy-adapter/src/main/java/com/airbnb/epoxy/ControllerModelList.kt
package com.airbnb.epoxy

/**
 * This ArrayList subclass enforces that no changes are made to the list after [.freeze] is
 * called. This prevents model interceptors from storing the list and trying to change it later. We
 * could copy the list before diffing, but that would waste memory to make the copy for every
 * buildModels cycle, plus the interceptors could still try to modify the list and be confused about
 * why it doesn't do anything.
 */
internal class ControllerModelList(expectedModelCount: Int) : ModelList(expectedModelCount) {
    companion object {
        private val OBSERVER: ModelListObserver = object : ModelListObserver {
            override fun onItemRangeInserted(positionStart: Int, itemCount: Int) {
                throw IllegalStateException("Models cannot be changed once they are added to the controller")
            }

            override fun onItemRangeRemoved(positionStart: Int, itemCount: Int) {
                throw IllegalStateException("Models cannot be changed once they are added to the controller")
            }
        }
    }

    init {
        pauseNotifications()
    }

    fun freeze() {
        setObserver(OBSERVER)
        resumeNotifications()
    }
}
